File: 1 - Arrays and Strings/OneOne.java
/*
Problem 1.1 - Cracking the Coding Interview

Implement an algorithm to determine if a string has all unique characters.
What if you cannot use additional data structures?

James Earle - August 26, 2015
*/
import java.io.*;

public class OneOne {
	
	public static final int ASCIIOFFSET = 65;
	BufferedReader reader;

	public OneOne() {
		reader = new BufferedReader(new InputStreamReader(System.in));
		try {
			String input = reader.readLine();
			System.out.println(hasUniqueChars(input.toUpperCase()));
		} catch(IOException e) {
			e.printStackTrace();
		}
	}

	public boolean hasUniqueChars(String input) {
		int[] arr = new int[26];
		String[] words = input.split(" ");

		for(String s : words) {
			for(int i=0;i<s.length();i++) {
				arr[s.charAt(i) - ASCIIOFFSET]++;
				if(arr[s.charAt(i) - ASCIIOFFSET] > 1) {
					System.out.println(s.charAt(i));
					return false;
				}	
			}
		}

		return true;

	}

	public static void main(String[] args) {
		new OneOne();
	}

}

File: C++/OneOne.cpp
#include <iostream>
#include <vector>

using namespace std;

/*
  Write an algorithm to determine if a string has all unique
  characters. What if you cannot use any additional data
  structures?
*/

int checkVector(vector<char> input);

int main()
{

  cout << "Input a word:" << endl;
  
  bool letters[128];
  
  vector<char> input;
  char c;
  
  while(cin >> c) {
    if(c == '\n') break;
    input.push_back(c);
  }
  
  return checkVector(input);
}

int checkVector(vector<char> input)
{
  // Set default to false.
  bool letters[128] = { false };
  
  for(int i=0;i<input.size();i++) {
    if(letters[input.at(i)]) {
      cout << input.at(i) << endl;
      cout << letters[input.at(i)] << endl;
      cout << "Word failed test" << endl;
      return 1;
    }
    
    // We've encountered this letter now.
    letters[input.at(i)] = true;
    
    // cout << "(" << input.at(i) << ", " << letters[input.at(i)] << ")" << endl;
  }
  
  cout << "Word has unique characters!" << endl;
  
  return 0;
}
